Journal attempts to read truncated entries

**Describe the bug**

When truncating entries, the journal will seek to the position we wish to truncate from and zero out the first 8 bytes (for length and checksum) to invalidate this entry. This, by itself, prevents readers from iterating past this, and is quite a cheap way of doing truncation.

However, imagine this case:

- A log with 4 entries: `{1},{2},{3}`
- We truncate entries `{2},{3}`, leaving only `{1}` - at this point we've zero-ed out the first 8 bytes after `{1}`.
- Say we append again another entry `{2}` with length less than the length of the previous `{2}` and `{3}` combined, e.g. such that some garbage from `{3}` is left behind
- When reading, the reader will read 1, 2, and then think there's more to read but it's all garbage data, which leads to errors.

**To Reproduce**

```java
/*
 * Copyright © 2017 camunda services GmbH (info@camunda.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.zeebe.journal.file;

import static org.assertj.core.api.Assertions.assertThat;

import io.atomix.utils.serializer.Namespace;
import io.atomix.utils.serializer.Namespaces;
import io.zeebe.journal.Journal;
import io.zeebe.journal.JournalReader;
import io.zeebe.journal.file.record.PersistedJournalRecord;
import java.nio.file.Path;
import org.agrona.DirectBuffer;
import org.agrona.concurrent.UnsafeBuffer;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class JournalTest {

  @TempDir Path directory;
  private final Namespace namespace =
      new Namespace.Builder()
          .register(Namespaces.BASIC)
          .nextId(Namespaces.BEGIN_USER_CUSTOM_ID)
          .register(PersistedJournalRecord.class)
          .register(UnsafeBuffer.class)
          .name("Journal")
          .build();
  private byte[] entry;
  private final DirectBuffer data = new UnsafeBuffer();
  private Journal journal;

  @BeforeEach
  void setup() {
    entry = "TestData".getBytes();
    data.wrap(entry);

    final int entrySize = getSerializedSize(data);
    final int entriesPerSegment = 10;
    journal = openJournal(entrySize, entriesPerSegment);
  }

  @Test
  void shouldNotReadDeletedEntries() {
    // given
    final var firstRecord = journal.append(data);
    journal.append(data);
    journal.append(data);

    // when
    journal.deleteAfter(firstRecord.index());
    final var newSecondRecord = journal.append(data);

    // then
    final JournalReader reader = journal.openReader();
    assertThat(newSecondRecord.index()).isEqualTo(2);

    assertThat(reader.hasNext()).isTrue();
    assertThat(reader.next()).isEqualTo(firstRecord);

    assertThat(reader.hasNext()).isTrue();
    assertThat(reader.next()).isEqualTo(newSecondRecord);

    assertThat(reader.hasNext()).isFalse();
  }

  private int getSerializedSize(final DirectBuffer data) {
    return namespace.serialize(new PersistedJournalRecord(1, 1, Integer.MAX_VALUE, data)).length
        + Integer.BYTES;
  }

  private SegmentedJournal openJournal(final int entrySize, final int entriesPerSegment) {
    return SegmentedJournal.builder()
        .withDirectory(directory.resolve("data").toFile())
        .withMaxSegmentSize(entriesPerSegment * entrySize + JournalSegmentDescriptor.BYTES)
        .withJournalIndexDensity(5)
        .build();
  }
}
```

**Expected behavior**

Once truncated, we cannot read entries again. One simple solution is to invalidate the next entry on every write. A better, more permanent solution should be introduced when we move away from Kryo.

**Log/Stacktrace**

This error occurred in the 1.0.0-alpha2 load test with the following symptom:

<details><summary>Full Stacktrace</summary>
 <p>

```
12:57:07.579 [] [main] ERROR io.zeebe.broker.system - Bootstrap Broker-0 [6/14]: cluster services failed with unexpected exception.
java.util.concurrent.CompletionException: com.esotericsoftware.kryo.KryoException: Buffer underflow.
Serialization trace:
byteArray (org.agrona.concurrent.UnsafeBuffer)
data (io.zeebe.journal.file.record.PersistedJournalRecord)
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1081) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: com.esotericsoftware.kryo.KryoException: Buffer underflow.
Serialization trace:
byteArray (org.agrona.concurrent.UnsafeBuffer)
data (io.zeebe.journal.file.record.PersistedJournalRecord)
	at com.esotericsoftware.kryo.io.Input.require(Input.java:221) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.io.Input.readBytes(Input.java:378) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.io.Input.readBytes(Input.java:358) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ByteArraySerializer.read(DefaultArraySerializers.java:55) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ByteArraySerializer.read(DefaultArraySerializers.java:36) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.Kryo.readObjectOrNull(Kryo.java:787) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:134) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.CompatibleFieldSerializer.read(CompatibleFieldSerializer.java:187) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:732) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:125) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.serializers.CompatibleFieldSerializer.read(CompatibleFieldSerializer.java:187) ~[kryo-5.0.3.jar:?]
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:813) ~[kryo-5.0.3.jar:?]
	at io.atomix.utils.serializer.Namespace.deserialize(Namespace.java:189) ~[classes/:?]
	at io.zeebe.journal.file.record.KryoSerializer.read(KryoSerializer.java:37) ~[classes/:?]
	at io.zeebe.journal.file.record.JournalRecordReaderUtil.read(JournalRecordReaderUtil.java:60) ~[classes/:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:195) ~[classes/:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:58) ~[classes/:?]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:149) ~[classes/:?]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:52) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:441) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:452) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:500) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:206) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:79) ~[classes/:?]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:136) ~[classes/:?]
	at io.atomix.raft.storage.log.RaftLog$Builder.build(RaftLog.java:320) ~[classes/:?]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:243) ~[classes/:?]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[classes/:?]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[classes/:?]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[classes/:?]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[classes/:?]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[classes/:?]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[classes/:?]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:119) ~[classes/:?]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[classes/:?]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(ConcurrentHashMap.java:3605) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.start(DefaultPartitionService.java:81) ~[classes/:?]
	at io.atomix.core.Atomix.lambda$startServices$0(Atomix.java:178) ~[classes/:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
	... 8 more
```

</p>
</details>


**Environment:**
 - OS: official docker image
 - Zeebe Version: 1.0.0-alpha2
 - Configuration: benchmark config

--- a/journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java
+++ b/journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java
@@ -26,7 +26,6 @@ import io.zeebe.journal.file.record.KryoSerializer;
 import io.zeebe.journal.file.record.PersistedJournalRecord;
 import java.nio.BufferOverflowException;
 import java.nio.BufferUnderflowException;
-import java.nio.ByteBuffer;
 import java.nio.MappedByteBuffer;
 import org.agrona.DirectBuffer;
 import org.agrona.IoUtil;
@@ -89,7 +88,7 @@ class MappedJournalSegmentWriter {
     // TODO: Should reject append if the asqn is not greater than the previous record
 
     final int recordStartPosition = buffer.position();
-    lastEntry = write(buffer, recordIndex, asqn, data);
+    lastEntry = write(recordIndex, asqn, data);
     index.index(lastEntry, recordStartPosition);
     return lastEntry;
   }
@@ -106,7 +105,7 @@ class MappedJournalSegmentWriter {
     }
 
     final int recordStartPosition = buffer.position();
-    lastEntry = write(buffer, record);
+    lastEntry = write(record);
     index.index(lastEntry, recordStartPosition);
   }
 
@@ -115,8 +114,7 @@ class MappedJournalSegmentWriter {
    * the method returns, the position of buffer will be advanced to a position were the next record
    * will be written.
    */
-  private JournalRecord write(
-      final ByteBuffer buffer, final long index, final long asqn, final DirectBuffer data) {
+  private JournalRecord write(final long index, final long asqn, final DirectBuffer data) {
 
     // compute checksum and construct the record
     // TODO: checksum should also include asqn. https://github.com/zeebe-io/zeebe/issues/6218
@@ -125,7 +123,7 @@ class MappedJournalSegmentWriter {
     final var checksum = checksumGenerator.compute(data);
     final var recordToWrite = new PersistedJournalRecord(index, asqn, checksum, data);
 
-    writeInternal(buffer, recordToWrite);
+    writeInternal(recordToWrite);
     return recordToWrite;
   }
 
@@ -133,16 +131,16 @@ class MappedJournalSegmentWriter {
    * Write the record to the buffer. After the method returns, the position of buffer will be
    * advanced to a position were the next record will be written.
    */
-  private JournalRecord write(final ByteBuffer buffer, final JournalRecord record) {
+  private JournalRecord write(final JournalRecord record) {
     final var checksum = checksumGenerator.compute(record.data());
     if (checksum != record.checksum()) {
       throw new InvalidChecksum("Checksum invalid for record " + record);
     }
-    writeInternal(buffer, record);
+    writeInternal(record);
     return record;
   }
 
-  private void writeInternal(final ByteBuffer buffer, final JournalRecord recordToWrite) {
+  private void writeInternal(final JournalRecord recordToWrite) {
     final int recordStartPosition = buffer.position();
     buffer.mark();
     if (recordStartPosition + Integer.BYTES > buffer.limit()) {
@@ -168,9 +166,22 @@ class MappedJournalSegmentWriter {
           "Entry size " + length + " exceeds maximum allowed bytes (" + maxEntrySize + ")");
     }
 
+    // invalidate next entry
+    final int nextEntryOffset = recordStartPosition + Integer.BYTES + length;
+    invalidateNextEntry(nextEntryOffset);
+
     buffer.position(recordStartPosition);
     buffer.putInt(length);
-    buffer.position(recordStartPosition + Integer.BYTES + length);
+    buffer.position(nextEntryOffset);
+  }
+
+  private void invalidateNextEntry(final int position) {
+    if (position + (Integer.BYTES * 2) >= buffer.capacity()) {
+      return;
+    }
+
+    buffer.putInt(position, 0);
+    buffer.putInt(position + Integer.BYTES, 0);
   }
 
   private void reset(final long index) {
@@ -210,18 +221,10 @@ class MappedJournalSegmentWriter {
 
     if (index < segment.index()) {
       buffer.position(JournalSegmentDescriptor.BYTES);
-      buffer.putInt(0);
-      buffer.putInt(0);
-      buffer.position(JournalSegmentDescriptor.BYTES);
+      invalidateNextEntry(JournalSegmentDescriptor.BYTES);
     } else {
-      // Reset the writer to the given index.
       reset(index);
-
-      // Zero entries after the given index.
-      final int position = buffer.position();
-      buffer.putInt(0);
-      buffer.putInt(0);
-      buffer.position(position);
+      invalidateNextEntry(buffer.position());
     }
   }
 

--- a/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
+++ b/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
@@ -35,7 +35,7 @@ import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
-public class JournalTest {
+class JournalTest {
 
   @TempDir Path directory;
   private final Namespace namespace =
@@ -51,7 +51,7 @@ public class JournalTest {
   private Journal journal;
 
   @BeforeEach
-  public void setup() {
+  void setup() {
     entry = "TestData".getBytes();
     data.wrap(entry);
 
@@ -61,13 +61,13 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldBeEmpty() {
+  void shouldBeEmpty() {
     // when-then
     assertThat(journal.isEmpty()).isTrue();
   }
 
   @Test
-  public void shouldNotBeEmpty() {
+  void shouldNotBeEmpty() {
     // given
     journal.append(1, data);
 
@@ -76,7 +76,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldAppendData() {
+  void shouldAppendData() {
     // when
     final var recordAppended = journal.append(1, data);
     assertThat(recordAppended.index()).isEqualTo(1);
@@ -88,7 +88,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldAppendMultipleRecords() {
+  void shouldAppendMultipleRecords() {
     // when
     for (int i = 0; i < 10; i++) {
       final var recordAppended = journal.append(i + 10, data);
@@ -109,7 +109,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldAppendAndReadMultipleRecords() {
+  void shouldAppendAndReadMultipleRecords() {
     final var reader = journal.openReader();
     for (int i = 0; i < 10; i++) {
       // given
@@ -132,7 +132,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldReset() {
+  void shouldReset() {
     // given
     long asqn = 1;
     assertThat(journal.getLastIndex()).isEqualTo(0);
@@ -150,7 +150,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldResetWhileReading() {
+  void shouldResetWhileReading() {
     // given
     final var reader = journal.openReader();
     long asqn = 1;
@@ -176,7 +176,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldWriteToTruncatedIndex() {
+  void shouldWriteToTruncatedIndex() {
     // given
     final var reader = journal.openReader();
     assertThat(journal.getLastIndex()).isEqualTo(0);
@@ -201,7 +201,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldTruncate() {
+  void shouldTruncate() {
     // given
     final var reader = journal.openReader();
     assertThat(journal.getLastIndex()).isEqualTo(0);
@@ -220,7 +220,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldNotReadTruncatedEntries() {
+  void shouldNotReadTruncatedEntries() {
     // given
     final int totalWrites = 10;
     final int truncateIndex = 5;
@@ -261,7 +261,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldAppendJournalRecord() {
+  void shouldAppendJournalRecord() {
     // given
     final var receiverJournal =
         SegmentedJournal.builder()
@@ -280,7 +280,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldNotAppendRecordWithAlreadyAppendedIndex() {
+  void shouldNotAppendRecordWithAlreadyAppendedIndex() {
     // given
     final var record = journal.append(1, data);
     journal.append(data);
@@ -290,7 +290,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldNotAppendRecordWithGapInIndex() {
+  void shouldNotAppendRecordWithGapInIndex() {
     // given
     final var receiverJournal =
         SegmentedJournal.builder()
@@ -305,7 +305,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldNotAppendLastRecord() {
+  void shouldNotAppendLastRecord() {
     // given
     final var record = journal.append(1, data);
 
@@ -314,7 +314,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldNotAppendRecordWithInvalidChecksum() {
+  void shouldNotAppendRecordWithInvalidChecksum() {
     // given
     final var receiverJournal =
         SegmentedJournal.builder()
@@ -333,7 +333,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldReturnFirstIndex() {
+  void shouldReturnFirstIndex() {
     // when
     final long firstIndex = journal.append(data).index();
     journal.append(data);
@@ -343,7 +343,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldReturnLastIndex() {
+  void shouldReturnLastIndex() {
     // when
     journal.append(data);
     final long lastIndex = journal.append(data).index();
@@ -353,7 +353,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldOpenAndClose() throws Exception {
+  void shouldOpenAndClose() throws Exception {
     // when/then
     assertThat(journal.isOpen()).isTrue();
     journal.close();
@@ -361,7 +361,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldReopenJournalWithExistingRecords() throws Exception {
+  void shouldReopenJournalWithExistingRecords() throws Exception {
     // given
     journal.append(data);
     journal.append(data);
@@ -378,7 +378,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldReadReopenedJournal() throws Exception {
+  void shouldReadReopenedJournal() throws Exception {
     // given
     final var appendedRecord = journal.append(data);
     journal.close();
@@ -394,7 +394,7 @@ public class JournalTest {
   }
 
   @Test
-  public void shouldWriteToReopenedJournalAtNextIndex() throws Exception {
+  void shouldWriteToReopenedJournalAtNextIndex() throws Exception {
     // given
     final var firstRecord = journal.append(data);
     journal.close();
@@ -413,6 +413,30 @@ public class JournalTest {
 
     assertThat(reader.hasNext()).isTrue();
     assertThat(reader.next()).isEqualTo(secondRecord);
+  }
+
+  @Test
+  void shouldNotReadDeletedEntries() {
+    // given
+    final var firstRecord = journal.append(data);
+    journal.append(data);
+    journal.append(data);
+
+    // when
+    journal.deleteAfter(firstRecord.index());
+    final var newSecondRecord = journal.append(data);
+
+    // then
+    final JournalReader reader = journal.openReader();
+    assertThat(newSecondRecord.index()).isEqualTo(2);
+
+    assertThat(reader.hasNext()).isTrue();
+    assertThat(reader.next()).isEqualTo(firstRecord);
+
+    assertThat(reader.hasNext()).isTrue();
+    assertThat(reader.next()).isEqualTo(newSecondRecord);
+
+    assertThat(reader.hasNext()).isFalse();
   }
 
   private int getSerializedSize(final DirectBuffer data) {

